File: platform/polyglot/raise_exceptions/py/src/common_exceptions/httpx/wrappers.py
"""
HTTPX error wrappers for common_exceptions.

Provides decorator and utilities to wrap HTTPX exceptions.
"""

import functools
from typing import Any, Callable, Optional, TypeVar, Union
from urllib.parse import urlparse

from ..base import BaseHttpException
from ..codes import ErrorCode
from ..inbound import BadRequestException
from ..internal import BadGatewayException
from ..logger import create
from ..outbound import (
    ConnectTimeoutException,
    NetworkException,
    ReadTimeoutException,
    UpstreamServiceException,
    UpstreamTimeoutException,
    WriteTimeoutException,
)

logger = create("common_exceptions", __file__)

# Type variable for generic function wrapping
F = TypeVar("F", bound=Callable[..., Any])


def extract_service_from_url(url: Union[str, Any]) -> str:
    """
    Extract service name from URL.

    Args:
        url: URL string or httpx URL object

    Returns:
        Service name (hostname)
    """
    try:
        url_str = str(url)
        parsed = urlparse(url_str)
        return parsed.netloc or parsed.path.split("/")[0] or "unknown"
    except Exception:
        return "unknown"


def httpx_error_to_exception(
    error: Exception,
    service: Optional[str] = None,
    operation: Optional[str] = None,
) -> BaseHttpException:
    """
    Convert an HTTPX exception to a common exception.

    Args:
        error: HTTPX exception
        service: Optional service name override
        operation: Optional operation name

    Returns:
        Appropriate BaseHttpException subclass
    """
    # Import httpx here to avoid hard dependency
    try:
        import httpx
    except ImportError:
        logger.error("httpx not installed, cannot convert error")
        return NetworkException(
            message=str(error),
            service=service,
            original_error=type(error).__name__,
        )

    error_type = type(error).__name__
    logger.debug(f"Converting HTTPX error: {error_type}")

    # Extract service from error if available
    if service is None and hasattr(error, "request") and error.request:
        service = extract_service_from_url(error.request.url)

    # Map HTTPX exceptions to common exceptions
    if isinstance(error, httpx.ConnectTimeout):
        return ConnectTimeoutException(
            service=service,
            timeout_ms=_extract_timeout(error),
        )

    if isinstance(error, httpx.ReadTimeout):
        return ReadTimeoutException(
            service=service,
            timeout_ms=_extract_timeout(error),
        )

    if isinstance(error, httpx.WriteTimeout):
        return WriteTimeoutException(
            service=service,
            timeout_ms=_extract_timeout(error),
        )

    if isinstance(error, httpx.PoolTimeout):
        # Pool exhaustion treated as connect timeout
        return ConnectTimeoutException(
            message="Connection pool exhausted",
            service=service,
            timeout_ms=_extract_timeout(error),
        )

    if isinstance(error, httpx.TimeoutException):
        # Generic timeout
        return UpstreamTimeoutException(
            service=service,
            operation=operation,
            timeout_ms=_extract_timeout(error),
        )

    if isinstance(error, httpx.HTTPStatusError):
        return UpstreamServiceException(
            service=service,
            operation=operation,
            upstream_status=error.response.status_code,
        )

    if isinstance(error, httpx.ConnectError):
        return NetworkException(
            message="Connection refused",
            service=service,
            original_error=str(error),
            code=ErrorCode.NETWORK_CONNECTION_REFUSED,
        )

    if isinstance(error, httpx.RemoteProtocolError):
        return BadGatewayException(
            message="Invalid response from upstream",
            details={"service": service, "error": str(error)} if service else None,
        )

    if isinstance(error, (httpx.InvalidURL, httpx.UnsupportedProtocol)):
        return BadRequestException(
            message=f"Invalid URL: {error}",
        )

    if isinstance(error, (httpx.ReadError, httpx.WriteError)):
        return NetworkException(
            message=str(error),
            service=service,
            original_error=error_type,
        )

    if isinstance(error, httpx.ProxyError):
        return NetworkException(
            message="Proxy error",
            service=service,
            original_error=str(error),
        )

    # Generic network error for other HTTPX errors
    if isinstance(error, (httpx.HTTPError, httpx.RequestError)):
        return NetworkException(
            message=str(error),
            service=service,
            original_error=error_type,
        )

    # Unknown error type
    return NetworkException(
        message=str(error),
        service=service,
        original_error=error_type,
    )


def _extract_timeout(error: Exception) -> Optional[int]:
    """Extract timeout value from error if available."""
    # HTTPX doesn't expose timeout value directly on exceptions
    # Return None to indicate unknown
    return None


def wrap_httpx_errors(
    service: Optional[str] = None,
    operation: Optional[str] = None,
) -> Callable[[F], F]:
    """
    Decorator to wrap HTTPX errors as common exceptions.

    Args:
        service: Optional service name (extracted from URL if not provided)
        operation: Optional operation name for context

    Returns:
        Decorated function that converts HTTPX errors

    Example:
        @wrap_httpx_errors(service="user-service")
        async def get_user(user_id: str):
            async with httpx.AsyncClient() as client:
                response = await client.get(f"/users/{user_id}")
                response.raise_for_status()
                return response.json()
    """
    try:
        import httpx
        httpx_available = True
    except ImportError:
        httpx_available = False

    def decorator(func: F) -> F:
        @functools.wraps(func)
        async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
            if not httpx_available:
                return await func(*args, **kwargs)

            try:
                return await func(*args, **kwargs)
            except httpx.HTTPError as e:
                logger.debug(f"Caught HTTPX error in {func.__name__}: {type(e).__name__}")
                raise httpx_error_to_exception(e, service, operation) from e

        @functools.wraps(func)
        def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
            if not httpx_available:
                return func(*args, **kwargs)

            try:
                return func(*args, **kwargs)
            except httpx.HTTPError as e:
                logger.debug(f"Caught HTTPX error in {func.__name__}: {type(e).__name__}")
                raise httpx_error_to_exception(e, service, operation) from e

        # Return appropriate wrapper based on function type
        import asyncio
        if asyncio.iscoroutinefunction(func):
            return async_wrapper  # type: ignore
        return sync_wrapper  # type: ignore

    return decorator


logger.debug("HTTPX wrappers initialized")

File: platform/polyglot/raise_exceptions/mjs/src/undici/wrappers.ts
/**
 * Undici error wrappers for common-exceptions.
 *
 * Provides utilities to wrap Undici/fetch exceptions.
 */

import { BaseHttpException } from '../base.js';
import { ErrorCode } from '../codes.js';
import { BadRequestException } from '../inbound.js';
import { InternalServerException, ServiceUnavailableException } from '../internal.js';
import {
  ConnectTimeoutException,
  NetworkException,
  ReadTimeoutException,
  UpstreamServiceException,
} from '../outbound.js';
import { create } from '../logger.js';

const logger = create('common-exceptions', __filename);

/**
 * Options for wrapUndiciErrors.
 */
export interface WrapUndiciErrorsOptions {
  service?: string;
  operation?: string;
}

/**
 * Extract service name from URL.
 */
export function extractServiceFromUrl(url: string | URL): string {
  try {
    const urlObj = typeof url === 'string' ? new URL(url) : url;
    return urlObj.hostname || 'unknown';
  } catch {
    return 'unknown';
  }
}

/**
 * Check if error is an Undici-specific error.
 */
function isUndiciError(error: unknown): boolean {
  if (!(error instanceof Error)) return false;

  const undiciErrorNames = [
    'UndiciError',
    'ConnectTimeoutError',
    'HeadersTimeoutError',
    'BodyTimeoutError',
    'SocketError',
    'RequestAbortedError',
    'InvalidArgumentError',
    'ClientClosedError',
    'ClientDestroyedError',
    'ResponseError',
  ];

  return undiciErrorNames.includes(error.name) || error.name.includes('Undici');
}

/**
 * Convert an Undici error to a common exception.
 *
 * @param error - Undici error or any error
 * @param service - Optional service name override
 * @param operation - Optional operation name
 * @returns Appropriate BaseHttpException subclass
 */
export function undiciErrorToException(
  error: Error,
  service?: string,
  _operation?: string
): BaseHttpException {
  const errorName = error.name;
  const errorMessage = error.message;

  logger.debug(`Converting Undici error: ${errorName}`);

  // Map Undici error types to common exceptions
  switch (errorName) {
    case 'ConnectTimeoutError':
      return new ConnectTimeoutException({ service });

    case 'HeadersTimeoutError':
    case 'BodyTimeoutError':
      return new ReadTimeoutException({ service });

    case 'SocketError':
      return new NetworkException({
        service,
        originalError: errorMessage,
      });

    case 'RequestAbortedError':
      return new NetworkException({
        message: 'Request was aborted',
        service,
        originalError: errorName,
        code: ErrorCode.NETWORK_ERROR,
      });

    case 'InvalidArgumentError':
      return new BadRequestException({
        message: `Invalid argument: ${errorMessage}`,
      });

    case 'ClientClosedError':
    case 'ClientDestroyedError':
      return new ServiceUnavailableException({
        message: 'HTTP client is closed',
      });

    default:
      // Check for common Node.js network errors
      if (errorMessage.includes('ECONNREFUSED')) {
        return new NetworkException({
          message: 'Connection refused',
          service,
          originalError: errorMessage,
          code: ErrorCode.NETWORK_CONNECTION_REFUSED,
        });
      }

      if (errorMessage.includes('ECONNRESET')) {
        return new NetworkException({
          message: 'Connection reset',
          service,
          originalError: errorMessage,
        });
      }

      if (errorMessage.includes('ENOTFOUND')) {
        return new NetworkException({
          message: 'DNS lookup failed',
          service,
          originalError: errorMessage,
          code: ErrorCode.NETWORK_DNS_FAILURE,
        });
      }

      if (errorMessage.includes('ETIMEDOUT')) {
        return new ConnectTimeoutException({ service });
      }

      // Generic network error for other cases
      return new NetworkException({
        message: errorMessage,
        service,
        originalError: errorName,
      });
  }
}

/**
 * Check upstream response status and throw if error.
 *
 * Undici/fetch doesn't auto-throw on 4xx/5xx, so this utility
 * checks the status and throws an UpstreamServiceException if needed.
 *
 * @param response - Fetch Response object
 * @param service - Service name for error context
 * @param operation - Optional operation name
 * @throws UpstreamServiceException if status >= 400
 */
export function checkUpstreamStatus(
  response: Response,
  service: string,
  operation?: string
): void {
  if (response.ok) return;

  logger.debug(`Upstream ${service} returned ${response.status}`);

  throw new UpstreamServiceException({
    service,
    operation,
    upstreamStatus: response.status,
  });
}

/**
 * Higher-order function to wrap async functions with Undici error handling.
 *
 * @param fn - Async function to wrap
 * @param options - Service and operation context
 * @returns Wrapped function that converts Undici errors
 *
 * @example
 * const fetchUser = wrapUndiciErrors(
 *   async (userId: string) => {
 *     const response = await fetch(`/users/${userId}`);
 *     checkUpstreamStatus(response, 'user-service');
 *     return response.json();
 *   },
 *   { service: 'user-service', operation: 'fetchUser' }
 * );
 */
export function wrapUndiciErrors<T extends (...args: any[]) => Promise<any>>(
  fn: T,
  options: WrapUndiciErrorsOptions = {}
): T {
  const { service, operation } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    try {
      return await fn(...args);
    } catch (error) {
      if (error instanceof BaseHttpException) {
        // Already a common exception, rethrow
        throw error;
      }

      if (error instanceof Error) {
        logger.debug(`Caught error in wrapped function: ${error.name}`);

        if (isUndiciError(error) || error.name === 'TypeError') {
          throw undiciErrorToException(error, service, operation);
        }

        // For other errors, wrap as network exception
        throw new NetworkException({
          message: error.message,
          service,
          originalError: error.name,
        });
      }

      // Unknown error type
      throw new InternalServerException({
        message: String(error),
      });
    }
  };

  return wrapped as T;
}

logger.debug('Undici wrappers initialized');
